librosa/feature/inverse.py

```diff
@@ -0,0 +1,263 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+'''Feature inversion'''
+
+import numpy as np
+import scipy.fftpack
+
+from ..core.spectrum import griffinlim
+from ..core.spectrum import db_to_power
+from .. import filters
+from ..util import nnls
+
+
+__all__ = ['mel_to_stft', 'mel_to_audio',
+           'mfcc_to_mel', 'mfcc_to_audio']
+
+
+def mel_to_stft(M, sr=22050, n_fft=2048, power=2.0, **kwargs):
+    '''Approximate STFT magnitude from a Mel power spectrogram.
+
+    Parameters
+    ----------
+    M : np.ndarray [shape=(n_mels, n), non-negative]
+        The spectrogram as produced by `feature.melspectrogram`
+
+    sr : number > 0 [scalar]
+        sampling rate of the underlying signal
+
+    n_fft : int > 0 [scalar]
+        number of FFT components in the resulting STFT
+
+    power : float > 0 [scalar]
+        Exponent for the magnitude melspectrogram
+
+    kwargs : additional keyword arguments
+        Mel filter bank parameters.
+        See `librosa.filters.mel` for details
+
+
+    Returns
+    -------
+    S : np.ndarray [shape=(n_fft, t), non-negative]
+        An approximate linear magnitude spectrogram
+
+
+    See Also
+    --------
+    feature.melspectrogram
+    core.stft
+    filters.mel
+    util.nnls
+
+
+    Examples
+    --------
+    >>> y, sr = librosa.load(librosa.util.example_audio_file(), duration=5, offset=10)
+    >>> S = np.abs(librosa.stft(y))
+    >>> mel_spec = librosa.feature.melspectrogram(S=S, sr=sr)
+    >>> S_inv = librosa.feature.inverse.mel_to_stft(mel_spec, sr=sr)
+
+    Compare the results visually
+
+    >>> import matplotlib.pyplot as plt
+    >>> plt.figure()
+    >>> plt.subplot(2,1,1)
+    >>> librosa.display.specshow(librosa.amplitude_to_db(S, ref=np.max, top_db=None),
+    ...                          y_axis='log', x_axis='time')
+    >>> plt.colorbar()
+    >>> plt.title('Original STFT')
+    >>> plt.subplot(2,1,2)
+    >>> librosa.display.specshow(librosa.amplitude_to_db(np.abs(S_inv - S),
+    ...                                                  ref=S.max(), top_db=None),
+    ...                          vmax=0, y_axis='log', x_axis='time', cmap='magma')
+    >>> plt.title('Residual error (dB)')
+    >>> plt.colorbar()
+    >>> plt.tight_layout()
+    '''
+
+    # Construct a mel basis with dtype matching the input data
+    mel_basis = filters.mel(sr, n_fft, n_mels=M.shape[0],
+                            dtype=M.dtype,
+                            **kwargs)
+
+    # Find the non-negative least squares solution, and apply
+    # the inverse exponent
+    return nnls(mel_basis, M)**(1./power)
```

Comment:
would it be worth to make a conditional branch for the case power=1? (and potentially power=2?) or is that a premature optimization?

Reply:
Last I checked, numpy is smart enough to special-case those inputs, so it shouldn't matter?  Or are you suggesting to do it in-place to save on memory copies?

Reply:
no, that's fine. i didn't know that numpy had optimization baked in.

Reply:
Ah, I spoke too soon.  Python optimizes these cases, but `np.power` does not.  So we can either get it fast (using python ops) or in-place (using `np.power`) but apparently not both.  Curious.

```python
In [11]: x = np.arange(int(1e6), dtype=np.float32)                                            

In [12]: %timeit np.power(x, 1.0)                                                             
11.4 ms ± 33.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [13]: %timeit np.power(x, 1)                                                               
11.6 ms ± 182 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [14]: %timeit np.power(x, 0.5)                                                             
11.3 ms ± 19.2 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [15]: np.power?                                                                            

In [16]: %timeit np.power(x, 2./3)                                                            
11.4 ms ± 7.19 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [17]: %timeit np.power(x, 1, out=x)                                                        
11.4 ms ± 55.6 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [18]: %timeit np.power(x, 1., out=x)                                                       
11.4 ms ± 46.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [19]: %timeit np.sqrt(x)                                                                   
342 µs ± 13.5 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [20]: %timeit x**(1./2)                                                                    
327 µs ± 5.39 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [21]: %timeit x**(1./1)                                                                    
498 µs ± 7.27 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```